### DISCUSSION TITLE
Missing required argument when using token_accessor

### DISCUSSION THREAD

**Discussion by olivierdewit** [Snowflake-Labs/terraform-provider-snowflake#2678] [Q&A] [answered]
Hi, 

I am using version 0.87.2 of the provider and I can't get a valid configuration when I use the token_accessor as recommended in the documentation. Here is the error I get when doing terraform validate:

Error: Missing required argument
│
│   with provider["registry.terraform.io/snowflake-labs/snowflake"],
│   on main.tf line 13, in provider "snowflake":
│   13: provider "snowflake" {
│
│ "oauth_client_secret": all of `oauth_client_id,oauth_client_secret,oauth_endpoint,oauth_redirect_url,oauth_refresh_token` must be specified

I can't find out what I am doing wrong. Here are my main.tf file and variables.tf:

main.tf:

terraform {
  required_version = ">= 1.1.7"
  backend "http" {
  }
  required_providers {
    snowflake = {
      source  = "Snowflake-Labs/snowflake"
      version = "0.87.2"
    }
  }
}

provider "snowflake" {
  account       = "XXX"
  user          = "SYSUSER_TF"
  authenticator = "TokenAccessor"
  token_accessor {
    client_id      = var.oauth_client_id
    client_secret  = var.oauth_client_secret
    redirect_uri   = var.oauth_redirect_url
    refresh_token  = var.oauth_refresh_token
    token_endpoint = var.oauth_endpoint
  }
}

variables.tf:

variable "oauth_client_secret" {
  type = string
}

variable "oauth_redirect_url" {
  type = string
}

variable "oauth_refresh_token" {
  type = string
}

variable "oauth_client_id" {
  type = string
}

variable "oauth_endpoint" {
  type = string
}

I even tried to change the authenticator to: authenticator = "OAuth". It doesn't change anything.

**Comment by sfc-gh-asawicki** [comment]
Do you have any of the `SNOWFLAKE_OAUTH_REFRESH_TOKEN`, `SNOWFLAKE_OAUTH_CLIENT_ID`, `SNOWFLAKE_OAUTH_CLIENT_SECRET`, `SNOWFLAKE_OAUTH_ENDPOINT`, or `SNOWFLAKE_OAUTH_REDIRECT_URL` env variables set?

**Reply by olivierdewit** [reply]
Since I declared the variables, I use instead TF_VAR_oauth_refresh_token, TF_VAR_oauth_client_id, TF_VAR_oauth_client_secret, TF_VAR_oauth_endpoint and TF_VAR_oauth_redirect_url as env variables. If I don't declare the variables, I get an error saying that my configuration is not valid.

**Reply by sfc-gh-asawicki** [reply]
I did not ask for the terraform variables but the specified environment variables (having at least one of them set would explain the error you mentioned, that's why I am asking for them first, to rule out this possibility).

**Reply by olivierdewit** [reply]
I see. Sorry for the confusion. In fact, yes SNOWFLAKE_OAUTH_REFRESH_TOKEN, SNOWFLAKE_OAUTH_CLIENT_ID, SNOWFLAKE_OAUTH_CLIENT_SECRET, SNOWFLAKE_OAUTH_ENDPOINT, and SNOWFLAKE_OAUTH_REDIRECT_URL are set as env variables. Should I remove them?

**Reply by sfc-gh-asawicki** [reply]
Yes, please. They are used to populate the old parameters, not the ones inside the `token_accessor` block.

**Reply by olivierdewit** [reply]
I did it but now I get the following error:

Error: could not retrieve access token from refresh token

**Reply by sfc-gh-asawicki** [reply]
Okay, so we are one step closer!

The error cause is unfortunately not propagated. I will change it, which will be included in the release this week. Until then, please run terraform with TF_LOG=DEBUG enabled and share them (please redact any sensitive information); I may be able to read something out of them.

**Reply by olivierdewit** [reply, marked as answer]
It is now working. The last error was due to the refresh token that had expired just this morning! Thank you for your help :-)

**Comment by qrious-kentran** [comment]
Hi,
just wonder what is your value for "oauth_redirect_url" ? I'm trying to authenticate to Snowflake using OAuth but couldn't find the value for the url. Thanks

**Reply by olivierdewit** [reply]
Hi,

oauth_redirect_url =  https://localhost.com
